File: common/src/main/java/com/study/common/security/InternalTokenProvider.java
package com.study.common.security;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.security.SecureRandom;
import java.time.Instant;
import java.util.Base64;

public class InternalTokenProvider {

    private static final String ALGORITHM = "AES/GCM/NoPadding";
    private static final int GCM_TAG_LENGTH = 128;
    private static final int IV_LENGTH = 12;
    private static final long TOKEN_VALIDITY_SECONDS = 30;

    private final SecretKeySpec keySpec;

    public InternalTokenProvider(String secretKey) {
        byte[] keyBytes = Base64.getDecoder().decode(secretKey);
        this.keySpec = new SecretKeySpec(keyBytes, "AES");
    }

    public String generateToken(String serviceName) {
        try {
            String payload = serviceName + ":" + Instant.now().getEpochSecond();

            byte[] iv = new byte[IV_LENGTH];
            new SecureRandom().nextBytes(iv);

            Cipher cipher = Cipher.getInstance(ALGORITHM);
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, new GCMParameterSpec(GCM_TAG_LENGTH, iv));
            byte[] encrypted = cipher.doFinal(payload.getBytes());

            ByteBuffer buffer = ByteBuffer.allocate(IV_LENGTH + encrypted.length);
            buffer.put(iv);
            buffer.put(encrypted);

            return Base64.getUrlEncoder().withoutPadding().encodeToString(buffer.array());
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate internal token", e);
        }
    }

    public TokenPayload validateToken(String token) {
        try {
            byte[] decoded = Base64.getUrlDecoder().decode(token);

            ByteBuffer buffer = ByteBuffer.wrap(decoded);
            byte[] iv = new byte[IV_LENGTH];
            buffer.get(iv);
            byte[] encrypted = new byte[buffer.remaining()];
            buffer.get(encrypted);

            Cipher cipher = Cipher.getInstance(ALGORITHM);
            cipher.init(Cipher.DECRYPT_MODE, keySpec, new GCMParameterSpec(GCM_TAG_LENGTH, iv));
            String payload = new String(cipher.doFinal(encrypted));

            String[] parts = payload.split(":");
            if (parts.length != 2) {
                return null;
            }

            String serviceName = parts[0];
            long timestamp = Long.parseLong(parts[1]);
            long now = Instant.now().getEpochSecond();

            if (Math.abs(now - timestamp) > TOKEN_VALIDITY_SECONDS) {
                return null;
            }

            return new TokenPayload(serviceName, timestamp);
        } catch (Exception e) {
            return null;
        }
    }

    public record TokenPayload(String serviceName, long timestamp) {}
}

File: common/src/main/kotlin/com/study/common/security/InternalTokenProvider.kt
package com.study.common.security

import java.nio.ByteBuffer
import java.security.SecureRandom
import java.time.Instant
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.math.abs

class InternalTokenProvider(secretKey: String) {

    private val keySpec: SecretKeySpec

    init {
        val keyBytes = Base64.getDecoder().decode(secretKey)
        keySpec = SecretKeySpec(keyBytes, "AES")
    }

    fun generateToken(serviceName: String): String {
        val payload = "$serviceName:${Instant.now().epochSecond}"
        val iv = ByteArray(IV_LENGTH).also { SecureRandom().nextBytes(it) }

        val cipher = Cipher.getInstance(ALGORITHM)
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, GCMParameterSpec(GCM_TAG_LENGTH, iv))
        val encrypted = cipher.doFinal(payload.toByteArray())

        val buffer = ByteBuffer.allocate(IV_LENGTH + encrypted.size)
        buffer.put(iv)
        buffer.put(encrypted)

        return Base64.getUrlEncoder().withoutPadding().encodeToString(buffer.array())
    }

    fun validateToken(token: String): TokenPayload? {
        return try {
            val decoded = Base64.getUrlDecoder().decode(token)
            val buffer = ByteBuffer.wrap(decoded)
            val iv = ByteArray(IV_LENGTH).also { buffer.get(it) }
            val encrypted = ByteArray(buffer.remaining()).also { buffer.get(it) }

            val cipher = Cipher.getInstance(ALGORITHM)
            cipher.init(Cipher.DECRYPT_MODE, keySpec, GCMParameterSpec(GCM_TAG_LENGTH, iv))
            val payload = String(cipher.doFinal(encrypted))

            val parts = payload.split(":")
            if (parts.size != 2) return null

            val serviceName = parts[0]
            val timestamp = parts[1].toLong()
            val now = Instant.now().epochSecond

            if (abs(now - timestamp) > TOKEN_VALIDITY_SECONDS) return null

            TokenPayload(serviceName, timestamp)
        } catch (e: Exception) {
            null
        }
    }

    data class TokenPayload(val serviceName: String, val timestamp: Long)

    companion object {
        private const val ALGORITHM = "AES/GCM/NoPadding"
        private const val GCM_TAG_LENGTH = 128
        private const val IV_LENGTH = 12
        private const val TOKEN_VALIDITY_SECONDS = 30L
    }
}
